sycl/tools/sycl-ls/sycl-ls.cpp

```diff
@@ -105,46 +111,129 @@ static void printSelectorChoice(const device_selector &Selector,
   if (filter) {
-    std::cerr << "Warning: SYCL_DEVICE_FILTER environment variable is set to "
-              << filter << "." << std::endl;
-    std::cerr << "To see device ids, please unset SYCL_DEVICE_FILTER."
-              << std::endl
-              << std::endl;
-    SuppressNumberPrinting = true;
+    if (!DiscardFilters) {
+      std::cerr << "INFO: Output filtered by SYCL_DEVICE_FILTER "
+                << "environment variable, which is set to " << filter << "."
+                << std::endl;
+      std::cerr << "To see device ids, use the --discard-filters CLI option."
+                << std::endl
+                << std::endl;
+      SuppressNumberPrinting = true;
+    } else
+      FilterEnvVars.insert({"SYCL_DEVICE_FILTER", filter});
   }
 #endif
 
   const char *ods_targets = std::getenv("ONEAPI_DEVICE_SELECTOR");
   if (ods_targets) {
-    std::cerr
-        << "Warning: ONEAPI_DEVICE_SELECTOR environment variable is set to "
-        << ods_targets << "." << std::endl;
-    std::cerr << "To see device ids, please unset ONEAPI_DEVICE_SELECTOR."
-              << std::endl
-              << std::endl;
-    SuppressNumberPrinting = true;
+    if (!DiscardFilters) {
+      std::cerr << "INFO: Output filtered by ONEAPI_DEVICE_SELECTOR "
+                << "environment variable, which is set to " << ods_targets
+                << "." << std::endl;
+      std::cerr << "To see device ids, use the --discard-filters CLI option."
+                << std::endl
+                << std::endl;
+      SuppressNumberPrinting = true;
+    } else
+      FilterEnvVars.insert({"ONEAPI_DEVICE_SELECTOR", ods_targets});
+  }
+
+  const char *sycl_dev_allow = std::getenv("SYCL_DEVICE_ALLOWLIST");
+  if (sycl_dev_allow) {
+    if (!DiscardFilters) {
+      std::cerr << "INFO: Output filtered by SYCL_DEVICE_ALLOWLIST "
+                << "environment variable, which is set to " << sycl_dev_allow
+                << "." << std::endl;
+      std::cerr << "To see device ids, use the --discard-filters CLI option."
+                << std::endl
+                << std::endl;
+      SuppressNumberPrinting = true;
+    } else
+      FilterEnvVars.insert({"SYCL_DEVICE_ALLOWLIST", sycl_dev_allow});
+  }
+}
+
+// Unset filter related environment variables namely, SYCL_DEVICE_FILTER,
+// ONEAPI_DEVICE_SELECTOR, and SYCL_DEVICE_ALLOWLIST.
+static void unsetFilterEnvVars() {
+  for (auto it : FilterEnvVars) {
+#ifdef _WIN32
+    _putenv_s(it.first.c_str(), "");
+#else
+    unsetenv(it.first.c_str());
+#endif
   }
+}
+
+// Restore filter related environment variables that we unset earlier.
+static void restoreFilterEnvVars() {
+  for (auto it : FilterEnvVars) {
+#ifdef _WIN32
+    _putenv_s(it.first.c_str(), it.second.c_str());
+#else
+    setenv(it.first.c_str(), it.second.c_str(), 1);
+#endif
+  }
+}
+
+int main(int argc, char **argv) {
+
+  if (argc == 1) {
+    verbose = false;
+    DiscardFilters = false;
+  } else {
```

Comment:
Do we use any command line options parsing libraries anywhere in sycl?

Reply:
We do use LLVM's CLI support (https://llvm.org/docs/CommandLine.html) at a couple of places but I'm not sure if we should use that here because we just have two CLI options for sycl-ls and both of them are just boolean flags.